lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -245,19 +300,242 @@ LogicalResult PipelineStageLowering::matchAndRewrite(
+  void updateInstance(RTLExternModuleOp mod, InstanceOp inst);
+  ESIRTLBuilder *B;
+};
+} // anonymous namespace
+
+/// Iterate through the `rtl.externmodule`s and lower their ports.
+void ESIPortsPass::runOnOperation() {
+  ModuleOp top = getOperation();
+  ESIRTLBuilder b(top);
+  B = &b;
+
+  for (auto mod : top.getOps<RTLExternModuleOp>()) {
+    updateFunc(mod);
+  }
+
+  B = nullptr;
+}
+
+/// Convert all input and output ChannelPorts into SV Interfaces. For inputs,
+/// just switch the type to `ModportType`. For outputs, append a `ModportType`
+/// to the inputs and remove the output channel from the results. Then call the
+/// function which adapts the instances (the hard part).
+void ESIPortsPass::updateFunc(RTLExternModuleOp mod) {
+  auto *ctxt = &getContext();
+  auto funcType = mod.getType();
+
+  bool updated = false;
+
+  // Reconstruct the list of operand types, changing the type whenever an ESI
+  // port is found.
+  SmallVector<Type, 16> newArgTypes;
+  for (auto argTy : funcType.getInputs()) {
+    auto chanTy = argTy.dyn_cast<ChannelPort>();
+    if (!chanTy) {
+      newArgTypes.push_back(argTy);
+      continue;
+    }
+
+    // When we find one, construct an interface, and add the 'sink' modport to
+    // the type list.
+    auto iface = B->getOrConstructInterface(chanTy);
+    newArgTypes.push_back(iface.getModportType(B->sink));
+    updated = true;
+  }
+
+  SmallVector<MutableDictionaryAttr, 16> argAttrs;
+  mod.getAllArgAttrs(argAttrs);
+  SmallVector<MutableDictionaryAttr, 16> resAttrs;
+  mod.getAllResultAttrs(resAttrs);
+
+  // Iterate through the results and append to one of the two below lists. The
+  // first for non-ESI-ports. The second, ports which have been re-located to an
+  // operand.
+  SmallVector<Type, 8> newResultTypes;
+  SmallVector<MutableDictionaryAttr, 4> newResultAttrs;
+  for (size_t resNum = 0, numRes = funcType.getNumResults(); resNum < numRes;
+       ++resNum) {
+    Type resTy = funcType.getResult(resNum);
+    auto chanTy = resTy.dyn_cast<ChannelPort>();
+    if (!chanTy) {
+      newResultTypes.push_back(resTy);
+      newResultAttrs.push_back(resAttrs[resNum]);
+      continue;
+    }
+
+    // When we find one, construct an interface, and add the 'sink' modport to
+    // the type list.
+    InterfaceOp iface = B->getOrConstructInterface(chanTy);
+    ModportType sourcePort = iface.getModportType(B->source);
+    newArgTypes.push_back(sourcePort);
+    argAttrs.push_back(resAttrs[resNum]);
+    updated = true;
+  }
+
+  if (!updated)
+    return;
+
+  // Set the new types.
+  auto newFuncType = FunctionType::get(newArgTypes, newResultTypes, ctxt);
+  mod.setType(newFuncType);
+  mod.setAllArgAttrs(argAttrs);
+  mod.setAllResultAttrs(newResultAttrs);
+
+  // Find all instances and update them.
+  Operation *scope = SymbolTable::getNearestSymbolTable(mod);
+  llvm::Optional<SymbolTable::UseRange> instances =
+      SymbolTable::getSymbolUses(mod, scope);
```

Comment:
Just FYI but getSymbolUses is an expensive function that leads to a full IR walk, is it the right way to go here?

Reply:
Is there a more efficient way to locate all of the instances of this module? They all need to be modified.

Reply:
Right, but you don't want to walk the entire module for each `RTLExternModuleOp`, that is N^2 algorithm.

I would recommend pulling restructuring your pass into two linear sweeps over the IR.

Reply:
Yeah, that made sense. I was just hoping there'd be some call which had a lookup cache or something like that to avoid a full IR walk. Guess that doesn't make much sense while the IR can be mutated.